homeassistant/components/dlib_face_identify/image_processing.py

```diff
@@ -14,44 +17,126 @@
+        vol.Required(CONF_FACES): vol.Any({cv.string: cv.isfile}, cv.isdir),
         vol.Optional(CONF_CONFIDENCE, default=0.6): vol.Coerce(float),
     }
 )
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up the Dlib Face detection platform."""
     entities = []
+
+    face_encodings_task = hass.async_create_task(
+        async_generate_encodings(hass, config[CONF_FACES])
+    )
+
     for camera in config[CONF_SOURCE]:
-        entities.append(
-            DlibFaceIdentifyEntity(
-                camera[CONF_ENTITY_ID],
-                config[CONF_FACES],
-                camera.get(CONF_NAME),
-                config[CONF_CONFIDENCE],
-            )
+
+        entity = DlibFaceIdentifyEntity(
+            camera[CONF_ENTITY_ID],
+            face_encodings_task,
+            camera.get(CONF_NAME),
+            config[CONF_CONFIDENCE],
         )
 
-    add_entities(entities)
+        entities.append(entity)
+
+    async_add_entities(entities)
+
+
+async def async_generate_encodings(hass, faces):
+    """Generate face encodings."""
+    face_encodings = {}
+
+    if not isinstance(faces, dict):
+        faces_root_path = pathlib.Path(faces)
+        if not faces_root_path.is_dir():
+            _LOGGER.error(
+                "Path '%s' does not exist or is not a valid directory. No face encodings generated.",
+                faces_root_path.resolve(),
+            )
+            return face_encodings
+
+        face_directories = [
+            faces_root_path / n for n in faces_root_path.iterdir() if n.is_dir()
+        ]
+
+        for face_images in face_directories:
+            face_encodings[face_images.stem] = []
+
+            for face_file in [
+                face_images / i for i in face_images.iterdir() if i.is_file()
+            ]:
+                if imghdr.what(face_file) is None:
+                    continue
+
+                try:
+                    image = await hass.async_add_executor_job(
+                        face_recognition.load_image_file, face_file
+                    )
+                    encodings_list = await hass.async_add_executor_job(
+                        face_recognition.face_encodings, image
+                    )
+
+                    if len(encodings_list) > 1:
+                        _LOGGER.error(
+                            "Failed to parse %s. More than one face detected in image.",
+                            face_file,
+                        )
+                    else:
+                        face_encodings[face_images.stem].append(encodings_list[0])
+
+                except IndexError as err:
+                    _LOGGER.error("Failed to parse %s. Error: %s", face_file, err)
+
+    else:
+
+        for face_name, face_file in faces.items():
+            try:
+                image = await hass.async_add_executor_job(
```

Comment:
Can we break out this part until the possible error log to a helper method that we can use here and above? Have the helper accept the face filename and return the first item of the  encodings_list. On error return `None`.

Reply:
Yeah, sure! Restructured code as requested in latest commit.